src/pow.rs

```diff
@@ -0,0 +1,1540 @@
+        }
+        Uint256(ret)
+    }
+
+    /// Wrapping (modular) addition. Computes `self + rhs`, wrapping around at the boundary of the
+    /// type.
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    fn wrapping_add(self, rhs: Self) -> Self {
+        let Uint256(ref me) = self;
+        let Uint256(ref you) = rhs;
+
+        let mut ret = [0u128; 2];
+        let mut carry = [0u128; 2];
+        let mut b_carry = false;
+        for i in (0..2).rev() {
+            let (res, overflow) = me[i].overflowing_add(you[i]);
+            ret[i] = res;
+            if i > 0 && overflow {
+                carry[i - 1] = 1;
+                b_carry = true;
+            }
+        }
+        if b_carry {
+            Uint256(ret).wrapping_add(Uint256(carry))
+        } else {
+            Uint256(ret)
+        }
+    }
+
+    /// Wrapping (modular) subtraction. Computes `self - rhs`, wrapping around at the boundary of
+    /// the type.
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    fn wrapping_sub(self, rhs: Self) -> Self { self.wrapping_add(!rhs).wrapping_add(Self::ONE) }
+
+    /// Wrapping (modular) multiplication. Computes `self * rhs`, wrapping around at the boundary of
+    /// the type.
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    fn wrapping_mul(self, rhs: Self) -> Self {
+        let mut me = Uint256::default();
+        for i in 0..4 {
+            let to_mul = (rhs >> (64 * i)).low_u64();
+            me = me.wrapping_add(self.mul_u64(to_mul) << (64 * i));
+        }
+        me
+    }
+
+    /// Panic-free bitwise shift-left; yields `self << mask(rhs)`, where `mask` removes any
+    /// high-order bits of `rhs` that would cause the shift to exceed the bitwidth of the type.
+    ///
+    /// Note that this is *not* the same as a rotate-left; the RHS of a wrapping shift-left is
+    /// restricted to the range of the type, rather than the bits shifted out of the LHS being
+    /// returned to the other end. We do not currently support `rotate_left`.
+    #[must_use = "this returns the result of the operation, without modifying the original"]
+    fn wrapping_shl(self, rhs: u32) -> Self {
+        let shift = rhs & 0x000000ff;
+
+        let Uint256(ref arr) = self;
+        let mut ret = [0u128; 2];
+        let word_shift = (shift / 128) as usize; // Max value is 3 because of mask above.
+        let bit_shift = shift % 128;
+
+        for (i, word) in arr.iter().enumerate().rev() {
+            let idx = i as i32 - word_shift as i32;
+            if idx >= 0 {
+                // Shift
+                ret[idx as usize] += word << bit_shift;
+                // Carry
+                let idx = idx - 1;
+                if bit_shift > 0 && idx >= 0 {
+                    ret[idx as usize] += word >> (128 - bit_shift);
+                }
+            }
+        }
```

Comment:
Now that we have two `u128` this can be simplified:

```rust
let left = self.0[0].wrapping_shl(shift) | self.0[1].wrapping_shr(128 - shift);
let right = self.0[1].wrapping_shl(shift);
Uint256([left, right])
```

Similarly below.